### PR TITLE
(SIMP-9033) Added .gitlab-ci.yml

### PR DESCRIPTION
**Pull request by billy-davies**
Adding .gitlab-ci.yml file with releng-checks
as well as the fingerprint rake task that
were present in the Travis CI tests.

[SIMP-9033] #close

[SIMP-9033]: https://simp-project.atlassian.net/browse/SIMP-9033

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  script:
+    - 'bundle exec rake fingerprint'
+
+releng_checks:
+  <<: *setup_bundler_env
+  stage: 'validation'
+  tags: ['docker']
+  script:
+    - 'if `hash apt-get`; then apt-get update; fi'
+    - 'if `hash apt-get`; then apt-get install -y rpm; fi'
```
Comment (op-ct) [inline .gitlab-ci.yml:82]: I suggest using the more idiomatic approach from the [standard module pipeline][0] to do this.  It avoids unnecessary updates/installs and forks:

```suggestion
    - 'command -v rpm || if command -v apt-get; then apt-get update; apt-get install -y rpm; fi ||:'
```

Also: **Beware of running commands in backticks inside ``if `<command>`; then``**.   The `if` statement will attempt to _execute the output_ of the forked `<command>` (which in the case of `hash` will be a blank string, unless if fails).

[0]: https://github.com/simp/pupmod-simp-simplib/blob/master/.gitlab-ci.yml#L295

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -0,0 +1,84 @@
+---
+stages:
+  - 'validation'
+
+variables:
+  PUPPET_VERSION:    'UNDEFINED' # <- Matrixed jobs MUST override this (or fail)
```
Comment (op-ct) [inline .gitlab-ci.yml:6]: > `Matrixed jobs MUST override this (or fail)`

This is true when `PUPPET_VERSION` is `UNDEFINED`

**Review by op-ct — CHANGES_REQUESTED** [review]

**Comment by op-ct** [comment]
The GLCI pipeline steps fail with `ERROR: Preparation failed: no Docker image specified to run the build in`, because docker jobs need an [`image`](https://docs.gitlab.com/ee/ci/yaml/#image) (see [example from our modules](https://github.com/simp/pupmod-simp-vsftpd/blob/master/.gitlab-ci.yml#L219))

File: .gitlab-ci.yml (hunk 2 above)
Comment (billy-davies) [reply .gitlab-ci.yml:6]: For the time being, I haven't put in any additional matrix data. I was going to suggest simply removing it but I haven't been able to see a gitlab-ci evaluate this change.

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+  SIMP_SKIP_NON_SIMPOS_TESTS: 1
+
+# bundler dependencies and caching
+#
+# - Cache bundler gems between pipelines foreach Ruby version
+# - Try to use cached and local resources before downloading dependencies
+# --------------------------------------
+.setup_bundler_env: &setup_bundler_env
+  image: 'ruby:2.5' # TODO: Default image for now, should add in images when this repository is ready for matrix testing.
```
Comment (op-ct) [inline .gitlab-ci.yml:25]: The jobs in this repository don't require matrix testing, so the repository can safely use a single ruby version to run its rake tasks.  For that reason, this `image:` setting would be best placed as a top-level key.

File: .gitlab-ci.yml (hunk 2 above)
Comment (op-ct) [reply .gitlab-ci.yml:6]: I'd remove `PUPPET_VERSION`; the Gemfile doesn't use it.

Before running in gitlab-ci, you can still test what will happen locally by setting the `PUPPET_VERSION` environment variable while running the steps in each job, e.g,:

```sh
export PUPPET_VERSION=UNDEFINED # or `unset PUPPET_VERSION`
bundle update
bundle exec rake fingerprint
```

**Pull request merged**